src: add web locks api

IlyasShabi opened the pull request:
This PR implements the [Web Locks API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Locks_API), Locks are used to coordinate access to shared resources across multiple threads.

This implementation is based on previous work in https://github.com/nodejs/node/pull/22719 and https://github.com/nodejs/node/pull/36502, but takes a C++ native approach for better performance and reliability, this solution uses a singleton `LockManager` with thread-safe data structures to coordinate locks across all workers.

- [x] Support `exclusive` and `shared` modes
- [x] Support `steal` option
- [x] Support `ifAvailable` option
- [x] Support `signal` option
- [x] Documentation
- [x] WPT tests
- [x] Add missing `query.https.any.js` tests as unit tests
- [x] Add basic tests

Closes: [#22702](https://github.com/nodejs/node/pull/36502)
Refs: https://w3c.github.io/web-locks

KhafraDev commented on lib/internal/locks.js:
```diff
@@ -0,0 +1,275 @@
+
+// https://w3c.github.io/web-locks/#api-lock
+class Lock {
+  constructor(...args) {
```
you have to do 

constructor(symbol = undefined, ...args) or else Lock.length will be 1 instead of 0

IlyasShabi replied:
WebIDL test is failing with `constructor(symbol, ...args)`, thanks @KhafraDev for your suggestion.
Done

atlowChemi commented on lib/internal/locks.js:
```diff
@@ -0,0 +1,275 @@
+          }
+        };
+
+        options.signal.addEventListener('abort', abortListener, { once: true });
```
This could use `addAbortListener` with `using`

IlyasShabi commented on lib/internal/locks.js:
```diff
@@ -0,0 +1,275 @@
+      steal: false,
+      signal: undefined,
+      ...options,
+    };
```
Im using WebIDL converters to validate options https://github.com/nodejs/node/pull/58666/files#diff-93449f67db6a581113b32306df634bd840b7793a736c2a2aef0a88a578e5e812R38

IlyasShabi commented on src/node_locks.h:
```diff
@@ -0,0 +1,142 @@
+
+class Lock final {
+ public:
+  enum Mode { kShared, kExclusive };
```
Done

IlyasShabi commented on lib/internal/navigator.js:
```diff
@@ -100,6 +103,14 @@ class Navigator {
+   * @return {LockManager}
+   */
+  get locks() {
+    this.#locks ??= locks;
```
Done

IlyasShabi commented on src/node_locks.h:
```diff
@@ -0,0 +1,142 @@
+
+namespace node {
+namespace worker {
+namespace locks {
```
Done

IlyasShabi commented on src/node_locks.cc:
```diff
@@ -0,0 +1,684 @@
+            // Immediately reject the stolen lock's released_promise
+            Local<Value> error =
+                Exception::Error(String::NewFromUtf8(isolate, kLockStolenError)
+                                     .ToLocalChecked());
```
Done

IlyasShabi commented on lib/internal/locks.js:
```diff
@@ -0,0 +1,275 @@
+    }
+
+    if (options.mode !== locks.LOCK_MODE_SHARED && options.mode !== locks.LOCK_MODE_EXCLUSIVE) {
+      throw new ERR_INVALID_ARG_VALUE('options.mode',
```
Not needed anymore

IlyasShabi commented on src/node_locks.cc:
```diff
@@ -0,0 +1,684 @@
+      if_available_(if_available) {
+  waiting_promise_.Reset(env_->isolate(), waiting);
+  released_promise_.Reset(env_->isolate(), released);
+  callback_.Reset(env_->isolate(), callback);
```
I added a test to verify this, could please check https://github.com/nodejs/node/pull/58666/files#diff-43b814c3b4b1a7ea09d2bf9c7b856da949bcb12366c4e64a705e0700a96595b3R144

IlyasShabi commented on lib/internal/locks.js:
```diff
@@ -0,0 +1,275 @@
+  validateAbortSignal,
+  validateDictionary,
+  validateFunction,
+  validateString,
```
Thanks for the suggestion, PR updated

mcollina reviewed (approved):
lgtm

panva commented:
@IlyasShabi can you rebase so that we can start CI?

BridgeAR commented:
There seems to be a related test failure on linuxone. I didn't look closer into the code, so I can't tell why: https://ci.nodejs.org/job/node-test-commit-linuxone/50100/

IlyasShabi commented:
@targos I remove it from globals, could you please replace `semver-major` by `semver-minor`

jasnell commented on lib/internal/locks.js:
```diff
@@ -0,0 +1,295 @@
+
+    if (signal) {
+      signal.throwIfAborted();
+    }
```
Nit: I'd suggest moving this up immediately after the `validateAbortSignal(...)` call ... or, move it into the `if (signal) {` block below.

jasnell commented on doc/api/worker_threads.md:
```diff
@@ -755,6 +755,143 @@ if (isMainThread) {
+process. The API mirrors the semantics of the
+[browser `LockManager`][]
+
+### Class: `Lock`
```
```suggestion
### Class: `worker.locks.Lock`
```

jasnell commented on doc/api/worker_threads.md:
```diff
@@ -755,6 +755,143 @@ if (isMainThread) {
+-->
+
+The `LockManager` interface provides methods for requesting and introspecting
+locks. To obtain a `LockManager` instance use `require('node:worker_threads').locks`.
```
I would put this in a code block sample that shows both ESM and CJS ways of accessing it

IlyasShabi commented on doc/api/worker_threads.md:
```diff
@@ -755,6 +755,143 @@ if (isMainThread) {
+
+The mode of the lock. Either `shared` or `exclusive`.
+
+### Class: `LockManager`
```
I don't think so since `worker.locks` is returning a `LockManager`

jasnell commented on src/node_locks.h:
```diff
@@ -0,0 +1,153 @@
+  mutable Mutex mutex_;
+  std::unordered_map<std::u16string, std::deque<std::shared_ptr<Lock>>>
+      held_locks_;
+  std::deque<std::unique_ptr<LockRequest>> pending_queue_;
```
`std::deque` is not the most efficient resource-wise. An `std::list` might be better here.

jasnell replied:
conceptually, yes, it is the better fit. std::deque, however, is quite a bit less efficient and uses more memory. Not a critical issue, however.

jasnell commented on src/node_locks.h:
```diff
@@ -0,0 +1,153 @@
+  std::unordered_map<std::u16string, std::deque<std::shared_ptr<Lock>>>
+      held_locks_;
+  std::deque<std::unique_ptr<LockRequest>> pending_queue_;
+  std::unordered_set<Environment*> registered_envs_;
```
This makes me nervous. We need to be certain that the pointers here are appropriately cleaned up so that if an `Environment*` gets destroyed it doesn't become a UAF. Can you add some documentation here pointing to where the cleanup happens so that others don't miss it.

jasnell commented on test/parallel/test-web-locks.js:
```diff
@@ -0,0 +1,176 @@
+      navigator.locks.request('exclusive-test', async (lock) => {
+        assert.strictEqual(lock.mode, 'exclusive');
+        parentPort.postMessage({ success: true });
+      }).catch(err => parentPort.postMessage({ error: err.message }));
```
It doesn't necessarily need to be done in this PR, but we should make sure that this plays well with AsyncLocalStorage. Specifically, something like the following should work:

```
const als = new AsyncLocalStorage();
als.run(123, () => {
  navigator.locks.request('exclusive-test', async (lock) => {
    assert.strictEqual(als.getStore(), 123);
  });
});
```

This would suggest that the lock request is capable of capturing the current async context frame and restoring it when the lock has been acquired and the callback is invoked.

jasnell commented on test/parallel/test-web-locks.js:
```diff
@@ -0,0 +1,176 @@
+      });
+    });
+
+    await worker.terminate();
```
There are a lot of tests here, which is excellent, but it makes it difficult to see if every case is covered. It would be good, if there isn't one already, to have a test that launches a worker thread that initiates a lock request but then have the worker get terminated/destroyed while the request is still pending to ensure that there aren't any UAFs.

IlyasShabi commented on src/node_locks.h:
```diff
@@ -0,0 +1,153 @@
+  Lock(const Lock&) = delete;
+  Lock& operator=(const Lock&) = delete;
+
+  const std::u16string& name() const { return name_; }
```
Done

jasnell commented on src/node_locks.cc:
```diff
@@ -0,0 +1,738 @@
+                       External::New(isolate, reject_holder))
+             .ToLocal(&on_rejected_callback)) {
+      delete fulfill_holder;
+      delete reject_holder;
```
These might need to be handled separately. If the first call succeeds, then the first created function will take ownership over the `fulfill_holder`. If the second call then fails for whatever reason, we are deleting the `fulfill_holder` while it's external is still holding the reference to it that it assumes it owns. It would be best to separate this into two separate calls rather than aggregating them together like this. Create one, create it's external and it's function, then create the second...

jasnell commented on src/node_locks.cc:
```diff
@@ -0,0 +1,738 @@
+        Local<Value> rejection_value = promise->Result();
+        grantable_request->waiting_promise()
+            ->Reject(context, rejection_value)
+            .Check();
```
Not something to do here, but using `Check()` here has the same issue as using `ToLocalChecked()` in that it will just crash the process rather than propagate the error. This is a common issue throughout the code, however, so not something I would block this PR on. We need to handle these better in general. 

If you did want to handle this here, then changing these to check if the return value is empty then doing some proper error propagation similar to the way the ToLocal(...) results are handled would be ideal.

jasnell commented on src/node_locks.cc:
```diff
@@ -0,0 +1,738 @@
+          ->Resolve(context, callback_result)
+          .Check();
+      Local<Value> promise_args[] = {callback_result};
+      USE(on_fulfilled_callback->Call(
```
Likewise here. Calling the callback could throw. The error should be appropriately propagated.

jasnell commented on src/node_locks.cc:
```diff
@@ -0,0 +1,738 @@
+    return;
+  }
+
+  args.GetReturnValue().Set(released_promise->GetPromise());
```
It's a bit unusual (by convention) to set the return value and then perform additional actions. It makes it rather easy to miss that the return value is set at all. Could this be moved to the end of the function?

IlyasShabi replied:
Done

jasnell commented on src/node_locks.cc:
```diff
@@ -0,0 +1,738 @@
+    for (const auto& resource_entry : manager->held_locks_) {
+      for (const auto& held_lock : resource_entry.second) {
+        if (held_lock->env() == env) {
+          Local<Object> lock_info =
```
You might consider moving the declaration for the `Local<Object> lock_info` to outside of the for loop and just reset it here so that we're reusing the same declaration on each iteration rather than creating a new one.

jasnell commented on src/node_locks.cc:
```diff
@@ -0,0 +1,738 @@
+    index = 0;
+    for (const auto& pending_request : manager->pending_queue_) {
+      if (pending_request->env() == env) {
+        Local<Object> lock_info =
```
Same here... if the `Local<Object> lock_info;` is moved outside of the for loops then it can just be reused here.

jasnell commented on src/node_locks.cc:
```diff
@@ -0,0 +1,738 @@
+    return Local<Object>();
+  }
+  obj->Set(context, FIXED_ONE_BYTE_STRING(isolate, "name"), name_string)
+      .Check();
```
Avoid Check here... and elsewhere in this function

jasnell commented on src/node_locks.cc:
```diff
@@ -0,0 +1,738 @@
+  LockManager::GetCurrent()->CleanupEnvironment(env);
+}
+
+static Local<Object> CreateLockInfoObject(Isolate* isolate,
```
Consider having this return a `MaybeLocal<Object>` to better facilitate error propagation

jasnell commented on src/node_locks.cc:
```diff
@@ -0,0 +1,738 @@
+  if (String::NewFromUtf8(isolate, kSharedMode).ToLocal(&shared_mode)) {
+    target->Set(FIXED_ONE_BYTE_STRING(isolate, "LOCK_MODE_SHARED"),
+                shared_mode);
+  }
```
Would `NODE_DEFINE_STRING_CONSTANT` work for these here instead?

IlyasShabi replied:
I don’t think so, because the `NODE_DEFINE_STRING_CONSTANT` macro is designed to work with `v8::Object`  not `v8::ObjectTemplate`

jasnell replied:
We should probably have it work with both at some point but 👍🏻

IlyasShabi commented:
@jasnell I fixed all your comments in the last commit, can you take a quick look and rerun the CI?
FYI a couple of WPT tests were flaky on linuxone RHEL 8/9 machines, but I cleaned up most of it 🤞

jasnell commented on src/node_locks.cc:
```diff
@@ -0,0 +1,901 @@
+                .ToLocalChecked();
+        Local<Value> err_val = Exception::Error(err_str_local);
+
+        USE(grantable_request->waiting_promise()->Reject(context, err_val));
+        USE(grantable_request->released_promise()->Reject(context, err_val));
```
We should not be using USE. These return a Maybe<bool> i believe, correct? If so, if those return an empty, then a error is scheduled and should be propagated. USE will cause that to be ignored.

jasnell replied:
In general yes, but if the first call to Reject fails we should early return rather than falling through to the next Reject.

IlyasShabi replied:
Yes I see, I pushed the suggestion

jasnell commented on src/node_locks.cc:
```diff
@@ -0,0 +1,901 @@
+    }
+  }
+
+  if (result->Set(context, FIXED_ONE_BYTE_STRING(isolate, "held"), held_list)
```
Nit: these strings can be added to `env_properties.h` instead of being created each time. Not something that needs to be done in this PR tho.

IlyasShabi replied:
Thanks for the suggestion, I did this for other strings too

IlyasShabi commented on src/node_locks.cc:
```diff
@@ -0,0 +1,738 @@
+  Local<Object> obj = Object::New(isolate);
+
+  Local<String> name_string;
+  if (!String::NewFromTwoByte(isolate,
```
Done

IlyasShabi replied:
I added a comment to address this later

IlyasShabi commented on src/node_locks.cc:
```diff
@@ -0,0 +1,901 @@
+          .IsNothing()) {
+    Local<String> error_msg =
+        String::NewFromUtf8(isolate, "Failed to build query result object")
+            .ToLocalChecked();
```
Done

IlyasShabi commented on test/wpt/status/web-locks.json:
```diff
@@ -0,0 +1,53 @@
+{
+  "held.https.any.js": {
```
Im marking this test as flaky since I saw node-test-commit-aix failing before. I’ll investigate and fix it in a separate PR

jasnell commented on src/env_properties.h:
```diff
@@ -96,6 +96,7 @@
   V(changes_string, "changes")                                                 \
   V(channel_string, "channel")                                                 \
   V(chunks_sent_since_last_write_string, "chunksSentSinceLastWrite")           \
+  V(clientId_string, "clientId")                                               \
```
For consistency, these should use lower-case, snake-case
```suggestion
  V(client_id_string, "clientId")                                              \
```

jasnell commented on src/node_locks.cc:
```diff
@@ -0,0 +1,880 @@
+
+// Called when the user callback promise rejects
+static void OnLockCallbackRejected(
+    const v8::FunctionCallbackInfo<v8::Value>& info) {
```
Here and other places ... 

```suggestion
    const FunctionCallbackInfo<Value>& info) {
```

IlyasShabi replied:
All `v8::…` have been replaced with their shorter aliases

jasnell commented on src/node_locks.cc:
```diff
@@ -0,0 +1,880 @@
+        // Then() failed, reject both promises and return
+        Local<String> err_str_local =
+            String::NewFromUtf8(isolate, "Failed to attach promise handlers")
+                .ToLocalChecked();
```
Use `FIXED_ONE_BYTE_STRING` instead?

IlyasShabi replied:
Done

jasnell commented on src/node_locks.cc:
```diff
@@ -0,0 +1,880 @@
+      Local<Promise> promise = callback_result.As<Promise>();
+      if (promise->Then(context, on_fulfilled_callback, on_rejected_callback)
+              .IsEmpty()) {
+        // Then() failed, reject both promises and return
```
If `Then` returns an empty, then v8 will have already scheduled an error. You can probably use a `v8::TryCatch` in here and reject the `waiting_promise()` and `released_promise()` with it's `Exception()`, but you'd want to check `CanContinue()` first.

IlyasShabi replied:
Done

jasnell reviewed (changes requested):
Just putting the red x on this since CI passed and it has sufficient sign off from others to land but there are still a few outstanding issues to resolve. I want to make sure it doesn't end up getting landed by the commit-queue while some comments are still pending.

IlyasShabi commented on src/node_locks.cc:
```diff
@@ -0,0 +1,880 @@
+                                                        callback_result));
+    } else {
+      USE(grantable_request->waiting_promise()->Resolve(context,
+                                                        callback_result));
```
Done

jasnell commented on src/node_locks.cc:
```diff
@@ -0,0 +1,934 @@
+// Called when the promise returned from the user's callback rejects
+static void OnIfAvailableReject(const FunctionCallbackInfo<Value>& info) {
+  HandleScope handle_scope(info.GetIsolate());
+  auto* holder = static_cast<Global<Promise::Resolver>*>(
```
These's aren't correct here. You really should not be holding a `v8::Global<T>` inside a `v8::External`, then deleting it like this. It's fine to use indirection through another type... e.g. having the `External` hold an instance of a struct that is holding the `Global<T>`.

IlyasShabi replied:
I did experiment with a wrapper that let the callback own a `std::unique_ptr`, so the `Global<…>` would be reset automatically, but that changed the timing of the reset and introduced a race that broke the WPT and hanging tests sometimes.

I think that the current pattern is also valid since the Global is reset and freed immediately in the callback and the raw pointer’s ownership is transferred with `release()`. Wdyt ?

jasnell replied:
I ought to be able to take another look this afternoon. Just catching up after two week of being out of the office ;-) ... just mentioning because I didn't want you to feel I was ignoring your pings on this ;-)

IlyasShabi replied:
Take your time, thanks for your help

addaleax replied:
Yeah, this pattern is indeed very odd. Here's a diff that simplifies things a lot without appearing to break anything:

```diff
@@ -132,24 +130,15 @@ static void OnLockCallbackRejected(const FunctionCallbackInfo<Value>& info) {
 // Called when the promise returned from the user's callback resolves
 static void OnIfAvailableFulfill(const FunctionCallbackInfo<Value>& info) {
   HandleScope handle_scope(info.GetIsolate());
-  auto* holder = static_cast<Global<Promise::Resolver>*>(
-      info.Data().As<External>()->Value());
-  USE(holder->Get(info.GetIsolate())
+  USE(info.Data().As<Promise::Resolver>()
           ->Resolve(info.GetIsolate()->GetCurrentContext(), info[0]));
-  holder->Reset();
-  delete holder;
 }
 
 // Called when the promise returned from the user's callback rejects
 static void OnIfAvailableReject(const FunctionCallbackInfo<Value>& info) {
   HandleScope handle_scope(info.GetIsolate());
-  auto* holder = static_cast<Global<Promise::Resolver>*>(
-      info.Data().As<External>()->Value());
-  USE(holder->Get(info.GetIsolate())
+  USE(info.Data().As<Promise::Resolver>()
           ->Reject(info.GetIsolate()->GetCurrentContext(), info[0]));
-  holder->Reset();
-
-  delete holder;
 }
 
 void LockManager::CleanupStolenLocks(Environment* env) {
@@ -326,28 +315,17 @@ void LockManager::ProcessQueue(Environment* env) {
       if (callback_result->IsPromise()) {
         Local<Promise> p = callback_result.As<Promise>();
 
-        // The resolver survives until the promise settles and is freed
-        // automatically.
-        auto resolve_holder = std::make_unique<Global<Promise::Resolver>>(
-            isolate, if_available_request->released_promise());
-        auto reject_holder = std::make_unique<Global<Promise::Resolver>>(
-            isolate, if_available_request->released_promise());
-
         Local<Function> on_fulfilled;
         Local<Function> on_rejected;
         CHECK(Function::New(context,
                             OnIfAvailableFulfill,
-                            External::New(isolate, resolve_holder.get()))
+                            if_available_request->released_promise())
                   .ToLocal(&on_fulfilled));
         CHECK(Function::New(context,
                             OnIfAvailableReject,
-                            External::New(isolate, reject_holder.get()))
+                            if_available_request->released_promise())
                   .ToLocal(&on_rejected));
 
-        // Transfer ownership to the callbacks.
-        resolve_holder.release();
-        reject_holder.release();
-
         {
           TryCatchScope try_catch_scope(env);
           if (p->Then(context, on_fulfilled, on_rejected).IsEmpty()) {
```

Generally speaking, we:
- Strongly want to avoid using `External`, because it is an "unsafe" feature of the V8 API in the sense that it associates a C++ object with a JS object without leaving V8's Garbage Collector, C++ memory management or diagnostic tooling any insight into what's happening (and so an `External` can easily end up accidentally containing a dangling pointer, or alternatively the C++ object not freed properly when the `External` is garbage-collected). 
- Avoid using `Global`s, because those do not properly represent references-relationships between objects (disclaimer: I don't know how the current work on C++ garbage collector integration changes this). If a `Global` pointing at a JS object B is contained in a C++ object that is attached to some JS object A, the garbage collector cannot see this reference, and it treats the Global as a separate GC root (or a weak reference), neither of which correctly represents the typical intended reference mechanism.

In this case, we should be able to just associate JS data directly with the JS function object in question. If this isn't possible, it's generally advisable to create a `BaseObject` with internal fields (in the V8 sense of the word) instead, which solves these issues.

I think the latter (inherit from `BaseObject`) is what we'd also want to do for `LockHolder`, instead of creating `External`s containing pointers to `LockHolder` instances.

jasnell commented:
I still think there may be at least a few issues here relating to how you're using `v8::Global` and `v8::External`. I'd really like to see if we could get @addaleax to take a look.

addaleax reviewed (changes requested):
Thank for the pings, in particular @BridgeAR for making me aware of this PR over Slack, where messages generally reach me a bit better 🙂 

This PR looks great and I am excited to see this feature finally come to Node.js, I think @jasnell had some good comments on the memory management aspects here that I think should be addressed before this PR is merged

addaleax commented:
(looks like CI failed because of a merge conflict? might just need to rebase this 🙂)

jasnell commented:
Landed in 062e8b5

joyeecheung commented:
`wpt/test-web-locks` has been flaky and failed at least 5 PRs out of the last 100 CI runs: https://github.com/nodejs/node/issues/59142

jasnell commented:
Yeah, been noticing that. Let's mark it flaky. @IlyasShabi do you think you'd be able to investigate?

IlyasShabi commented:
@jasnell  I have a PR to mark it flaky https://github.com/nodejs/node/pull/59144. I will do investigate this later

aduh95 commented:
This doesn't land cleanly on `v22.x-staging`, we would need a manual backport PR if we want it on `v22.x`.
